TorchServe on AKS (#644)
pytorch/serve: kubernetes/AKS/README.md

`az aks create  --resource-group myResourceGroup  --name myAKSCluster --node-vm-size Standard_NC6   --node-count 1`

#### 1.4 Connect to the cluster

To manage a Kubernetes cluster, you use [kubectl](https://kubernetes.io/docs/user-guide/kubectl/), the Kubernetes command-line client. If you use Azure Cloud Shell, `kubectl` is already installed. To install `kubectl` locally, use the [az aks install-cli](https://docs.microsoft.com/zh-cn/cli/azure/aks?view=azure-cli-latest#az-aks-install-cli) command:

`az aks install-cli`

To configure `kubectl` to connect to your Kubernetes cluster, use the [az aks get-credentials](https://docs.microsoft.com/zh-cn/cli/azure/aks?view=azure-cli-latest#az-aks-get-credentials) command. This command downloads credentials and configures the Kubernetes CLI to use them.

`az aks get-credentials --resource-group myResourceGroup --name myAKSCluster`

#### 1.5 Install NVIDIA device plugin

Before the GPUs in the nodes can be used, you must deploy a DaemonSet for the NVIDIA device plugin. This DaemonSet runs a pod on each node to provide the required drivers for the GPUs.

> `kubectl apply -f nvidia-device-plugin-ds.yaml`

Reviewer: Just curious. 

Can you use helm to install this as used in EKS.

```
helm repo add nvdp https://nvidia.github.io/k8s-device-plugin &&
helm repo update &&
helm install \
    --version=0.6.0 \
    --generate-name \
    nvdp/nvidia-device-plugin
```

Verify that the PVC / PV is created by excuting.

`kubectl get pvc,pv`

Your output should look similar to

NAME                                      STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS      AGE
persistentvolumeclaim/model-store-claim   Bound    pvc-c9e235a8-ca2b-4d04-8f25-8262de1bb915   1Gi        RWO            managed-premium   29s

NAME                                                        CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                       STORAGECLASS      REASON   AGE
persistentvolume/pvc-c9e235a8-ca2b-4d04-8f25-8262de1bb915   1Gi        RWO            Delete           Bound    default/model-store-claim   managed-premium            28s

#### 2.3 Create a pod and copy MAR / config files

Create a pod named `pod/model-store-pod` with PersistentVolume mounted so that we can copy the MAR / config files.

`kubectl apply -f pod.yaml`

Your output should look similar to

pod/model-store-pod created

Verify that the pod is created by excuting.

`kubectl get po`

Your output should look similar to

NAME                               READY   STATUS    RESTARTS   AGE

> model-store-pod                    1/1     Running   0          39s

Reviewer: Were you able to test this for a multi node / pod deployment. 

With a single node we many not be able to test if the shared volume is being used.

## TorchServe on Azure Kubernetes Service (AKS)

> ### 1、Create an AKS cluster

Reviewer 2: ```suggestion
### 1 Create an AKS cluster
```

> This quickstart requires that you are running the Azure CLI version 2.0.64 or later. Run `az --version` to find the version. If you need to install or upgrade, see [Install Azure CLI](https://docs.microsoft.com/zh-cn/cli/azure/install-azure-cli).

Reviewer 3: ```suggestion
This quickstart requires that you are running the Azure CLI version 2.0.64 or later. Run `az --version` to find the version. If you need to install or upgrade, see [Install Azure CLI](https://docs.microsoft.com/cli/azure/install-azure-cli?view=azure-cli-latest).
```

Reviewer 3: Remove reference to locale-specific documentation

```az aks create  --resource-group myResourceGroup  --name myAKSCluster --node-vm-size Standard_NC6   --node-count 1```

#### 1.4 Connect to the cluster

To manage a Kubernetes cluster, you use [kubectl](https://kubernetes.io/docs/user-guide/kubectl/), the Kubernetes command-line client. If you use Azure Cloud Shell, `kubectl` is already installed. To install `kubectl` locally, use the [az aks install-cli](https://docs.microsoft.com/en-us/cli/azure/aks?view=azure-cli-latest#az-aks-install-cli) command:

```az aks install-cli```

To configure `kubectl` to connect to your Kubernetes cluster, use the [az aks get-credentials](https://docs.microsoft.com/en-us/cli/azure/aks?view=azure-cli-latest#az-aks-get-credentials) command. This command downloads credentials and configures the Kubernetes CLI to use them.

```az aks get-credentials --resource-group myResourceGroup --name myAKSCluster```

#### 1.5 Install NVIDIA device plugin

Before the GPUs in the nodes can be used, you must deploy a DaemonSet for the NVIDIA device plugin. This DaemonSet runs a pod on each node to provide the required drivers for the GPUs.

> ```kubectl apply -f nvidia-device-plugin-ds.yaml```

Reviewer 3: We don't have the .yaml file yet in this step because we clone the repo in step 2.1. Can we change the order so that the repo is cloned earlier or the NVidia plugin is applied later?

Author: Thanks for your comments. The order of AKS/README.md has been changed.

NAME                  READY  STATUS  RESTARTS  AGE

nvidia-device-plugin-daemonset-7lvxd  1/1   Running  0     42s

#### 1.6 Install helm

```
curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3
chmod 700 get_helm.sh
./get_helm.sh
```

> ### 2、Deploy TorchServe on AKS

Reviewer 2: ```suggestion
### 2 Deploy TorchServe on AKS
```

```kubectl apply -f templates/nvidia-device-plugin-ds.yaml```

> 

Reviewer 3: nit:
```suggestion
`kubectl get pods` should show something similar to:
```

> NAME                  READY  STATUS  RESTARTS  AGE
> 
> nvidia-device-plugin-daemonset-7lvxd  1/1   Running  0     42s

Reviewer 3: nit:
```suggestion
NAME                  READY  STATUS  RESTARTS  AGE
nvidia-device-plugin-daemonset-7lvxd  1/1   Running  0     42s
```

> nvidia-device-plugin-daemonset-7lvxd  1/1   Running  0     42s

Reviewer 3: nit: Since the EKS document puts log results in boxes, let's do the same to keep the docs consistent.
(Note: I'm unable to actually suggest the change because of how markdown works)
```
NAME                  READY  STATUS  RESTARTS  AGE
nvidia-device-plugin-daemonset-7lvxd  1/1   Running  0     42s
```

#### 2.3 Create a storage class

A storage class is used to define how an Azure file share is created. If multiple pods need concurrent access to the same storage volume, you need Azure Files. Create the storage class with the following kubectl apply command:

```kubectl apply -f templates/Azure_file_sc.yaml```

#### 2.4 Create PersistentVolume

```kubectl apply -f templates/AKS_pv_claim.yaml```

Your output should look similar to

> persistentvolumeclaim/model-store-claim created

Reviewer 3: nit:
```suggestion
```persistentvolumeclaim/model-store-claim created```
```

> persistentvolumeclaim/model-store-claim created

Reviewer 3: nit:
```suggestion
```persistentvolumeclaim/model-store-claim created```
```

Verify that the PVC / PV is created by excuting.

```kubectl get pvc,pv```

Your output should look similar to

NAME                                      STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS      AGE
persistentvolumeclaim/model-store-claim   Bound    pvc-c9e235a8-ca2b-4d04-8f25-8262de1bb915   1Gi        RWO            managed-premium   29s

NAME                                                        CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                       STORAGECLASS      REASON   AGE

> persistentvolume/pvc-c9e235a8-ca2b-4d04-8f25-8262de1bb915   1Gi        RWO            Delete           Bound    default/model-store-claim   managed-premium            28s

Reviewer 3: nit: wrap in a box with ```

#### 2.4 Create a pod and copy MAR / config files

Create a pod named `pod/model-store-pod` with PersistentVolume mounted so that we can copy the MAR / config files.

> ```kubectl apply -f templates/model-store-pod.yaml```

Reviewer 3: ```suggestion
```kubectl apply -f templates/model_store_pod.yaml```
```
The checked in yaml file uses underscores instead

Your output should look similar to

> pod/model-store-pod created

Reviewer 3: ```suggestion
```pod/model-store-pod created```
```

Verify that the pod is created by excuting.

```kubectl get po```

Your output should look similar to

NAME                                   READY   STATUS    RESTARTS   AGE
model-store-pod                        1/1     Running   0          143m
nvidia-device-plugin-daemonset-qccgn   1/1     Running   0          144m

> torchserve-576df559ff-tww7q            1/1     Running   0          141m

Reviewer 3: nit: wrap in box with ```

> NAME                                   READY   STATUS    RESTARTS   AGE
> model-store-pod                        1/1     Running   0          143m
> nvidia-device-plugin-daemonset-qccgn   1/1     Running   0          144m
> torchserve-576df559ff-tww7q            1/1     Running   0          141m

Reviewer 3: torchserve pod not created yet in this step
```suggestion
NAME                                   READY   STATUS    RESTARTS   AGE
model-store-pod                        1/1     Running   0          143m
nvidia-device-plugin-daemonset-qccgn   1/1     Running   0          144m
```

#### 2.6 Down and copy MAR / config files

```shell

> wget https://torchserve.s3.amazonaws.com/mar_files/squeezenet1_1.mar
> wget https://torchserve.s3.amazonaws.com/mar_files/mnist.mar

Reviewer 3: Use new pytorch.org URLs for mar files
```suggestion
wget https://torchserve.pytorch.org/mar_files/squeezenet1_1.mar
wget https://torchserve.pytorch.org/mar_files/mnist.mar
```

kubectl exec --tty pod/model-store-pod -- mkdir /mnt/azure/model-store/
kubectl cp squeezenet1_1.mar model-store-pod:/mnt/azure/model-store/squeezenet1_1.mar
kubectl cp mnist.mar model-store-pod:/mnt/azure/model-store/mnist.mar

kubectl exec --tty pod/model-store-pod -- mkdir /mnt/azure/config/
kubectl cp config.properties model-store-pod:/mnt/azure/config/config.properties
```

Verify that the MAR / config files have been copied to the directory.

```kubectl exec --tty pod/model-store-pod -- find /mnt/azure/```

Your output should look similar to

/mnt/azure/
/mnt/azure/config
/mnt/azure/config/config.properties
/mnt/azure/lost+found
/mnt/azure/model-store
/mnt/azure/model-store/mnist.mar

> /mnt/azure/model-store/squeezenet1_1.mar

Reviewer 3: nit: wrap in ```

#### 2.7 Install Torchserve using Helm Charts

> Enter the Helm directory and install TorchServe using Helm Charts

Reviewer 3: nit:
```suggestion
Enter the Helm directory and install TorchServe using Helm Charts.
```

```cd ../Helm```

```helm install ts .```

Your output should look similar to

NAME: ts
LAST DEPLOYED: Thu Aug 20 02:07:38 2020
NAMESPACE: default
STATUS: deployed
REVISION: 1

> TEST SUITE: None

Reviewer 3: wrap in ```

#### 2.8 Check the status of TorchServe

```kubectl get po```

The installation will take a few minutes. Output like this means the installation is not completed yet.

NAME                               READY   STATUS              RESTARTS   AGE

> torchserve-75f5b67469-5hnsn        0/1     ContainerCreating   0          6s

Reviewer 3: wrap in ```

Output like this means the installation is completed.

NAME                               READY   STATUS    RESTARTS   AGE

> torchserve-75f5b67469-5hnsn        1/1     Running   0          2m36s

Reviewer 3: wrap in ```

### 3 Test Torchserve Installation

#### 3.1 Fetch the Load Balancer Extenal IP

Fetch the Load Balancer Extenal IP by executing.

```kubectl get svc```

Your output should look similar to

NAME               TYPE           CLUSTER-IP     EXTERNAL-IP    PORT(S)                         AGE
kubernetes         ClusterIP      10.0.0.1       <none>         443/TCP                         5d19h

> torchserve         LoadBalancer   10.0.39.88     your-external-IP   8080:30306/TCP,8081:30442/TCP   48s

Reviewer 3: wrap in ```
